File: TPs/TP2/Personagem.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

// Definição da classe Personagem
public class Personagem {
    private String id;
    private String name;
    private String alternateNames;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private boolean hogwartsStaff;
    private boolean hogwartsStudent;
    private String actorName;
    private boolean alive;
    private String alternateActors;
    private String dateOfBirth;
    private int yearOfBirth;
    private String eyeColour;
    private String gender;
    private String hairColour;
    private boolean wizard;

    // Lista para armazenar os objetos Personagem
    private static List<Personagem> listaPersonagens = new ArrayList<>();

    // Nome do arquivo CSV que contém os dados dos personagens
    private static final String FILE_NAME = "/tmp/characters.csv";

    // Construtor parametrizado para inicializar os atributos de Personagem
    public Personagem(String id, String name, String alternateNames, String house, String ancestry, String species,
                      String patronus, boolean hogwartsStaff, boolean hogwartsStudent, String actorName, boolean alive,
                      String alternateActors, String dateOfBirth, int yearOfBirth, String eyeColour, String gender,
                      String hairColour, boolean wizard) {
        this.id = id;
        this.name = name;
        this.alternateNames = alternateNames;
        this.house = house;
        this.ancestry = ancestry;
        this.species = species;
        this.patronus = patronus;
        this.hogwartsStaff = hogwartsStaff;
        this.hogwartsStudent = hogwartsStudent;
        this.actorName = actorName;
        this.alive = alive;
        this.alternateActors = alternateActors;
        this.dateOfBirth = dateOfBirth;
        this.yearOfBirth = yearOfBirth;
        this.eyeColour = eyeColour;
        this.gender = gender;
        this.hairColour = hairColour;
        this.wizard = wizard;
    }

    // Método imprimir para exibir os detalhes de um personagem
    public void imprimir() {
        System.out.print("[" + id);
        System.out.print(" ## " + name);
        System.out.print(" ## " + alternateNames);
        System.out.print(" ## " + house);
        System.out.print(" ## " + ancestry);
        System.out.print(" ## " + species);
        System.out.print(" ## " + patronus);
        System.out.print(" ## " + hogwartsStaff);
        System.out.print(" ## " + hogwartsStudent);
        System.out.print(" ## " + actorName);
        System.out.print(" ## " + alive);
        System.out.print(" ## " + alternateActors);
        System.out.print(" ## " + dateOfBirth);
        System.out.print(" ## " + yearOfBirth);
        System.out.print(" ## " + eyeColour);
        System.out.print(" ## " + gender);
        System.out.print(" ## " + hairColour);
        System.out.print(" ## " + wizard);
        System.out.println("]");
    }

    // Método estático para ler dados do arquivo CSV e preencher a lista de Personagem
    public static void lerDadosDoArquivo() {
        listaPersonagens.clear();
        try (BufferedReader ler = new BufferedReader(new FileReader(FILE_NAME))) {
            String ln;
            while ((ln = ler.readLine()) != null) {
                String[] atributo = ln.split(";");
                if (atributo.length == 18) { // Garante que todas as 18 colunas estão presentes
                    String id = atributo[0].trim();
                    String name = atributo[1].trim();
                    String alternateNames = atributo[2].trim();
                    String house = atributo[3].trim();
                    String ancestry = atributo[4].trim();
                    String species = atributo[5].trim();
                    String patronus = atributo[6].trim();
                    boolean hogwartsStaff = Boolean.parseBoolean(atributo[7].trim());
                    boolean hogwartsStudent = Boolean.parseBoolean(atributo[8].trim());
                    String actorName = atributo[9].trim();
                    boolean alive = Boolean.parseBoolean(atributo[10].trim());
                    String alternateActors = atributo[11].trim();
                    String dateOfBirth = atributo[12].trim();
                    int yearOfBirth = Integer.parseInt(atributo[13].trim());
                    String eyeColour = atributo[14].trim();
                    String gender = atributo[15].trim();
                    String hairColour = atributo[16].trim();
                    boolean wizard = Boolean.parseBoolean(atributo[17].trim());
    
                    Personagem personagem = new Personagem(id, name, alternateNames, house, ancestry, species, patronus,
                            hogwartsStaff, hogwartsStudent, actorName, alive, alternateActors, dateOfBirth, yearOfBirth,
                            eyeColour, gender, hairColour, wizard);
                    listaPersonagens.add(personagem);
                } else {
                    System.out.println("Linha mal formatada ou incompleta: " + ln);
                }
            }
        } catch (IOException x) {
            x.printStackTrace();
        }
    }

    // Método main
    public static void main(String[] args) {
        lerDadosDoArquivo(); // Chama o método para ler dados do arquivo CSV
        // Loop para criar objetos Personagem a partir da entrada do usuário
        while (true) {
            String input = MyIO.readLine().trim();
            if (input.equalsIgnoreCase("FIM")) {
                break;
            }
            String personagemId = input;
            boolean personagemEncontrado = false;

            for (Personagem personagem : listaPersonagens) {
                if (personagem.id.equals(personagemId)) {
                    personagemEncontrado = true;
                    personagem.imprimir();
                    break;
                }
            }
            if (!personagemEncontrado) {
                System.out.println("Personagem não encontrado");
            }
        }
    }
}

File: TPs/TP2/Personagem.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// Definindo a estrutura do Personagem
typedef struct {
    char id[100];
    char name[100];
    char alternateNames[500]; // Pode ser uma string com os nomes alternativos separados por vírgula
    char house[100];
    char ancestry[100];
    char species[100];
    char patronus[100];
    bool hogwartsStaff;
    bool hogwartsStudent;
    char actor[100];
    bool alive;
    char dateOfBirth[100];
    char yearOfBirth[100];
    char eyeColour[100];
    char gender[100];
    char hairColour[100];
    bool wizard;
} Character;

// Função para ler os dados de um personagem a partir de uma string
void readCharacterData(char* str, Character* character) {
    char* token = strsep(&str, ";");
    strcpy(character->id, token);

    token = strsep(&str, ";");
    strcpy(character->name, token);

    token = strsep(&str, ";");
    strcpy(character->alternateNames, token);

    token = strsep(&str, ";");
    strcpy(character->house, token);

    token = strsep(&str, ";");
    strcpy(character->ancestry, token);

    token = strsep(&str, ";");
    strcpy(character->species, token);

    token = strsep(&str, ";");
    strcpy(character->patronus, token);

    token = strsep(&str, ";");
    character->hogwartsStaff = (strcmp(token, "1") == 0);

    token = strsep(&str, ";");
    character->hogwartsStudent = (strcmp(token, "1") == 0);

    token = strsep(&str, ";");
    strcpy(character->actor, token);

    token = strsep(&str, ";");
    character->alive = (strcmp(token, "1") == 0);

    token = strsep(&str, ";");
    strcpy(character->dateOfBirth, token);

    token = strsep(&str, ";");
    strcpy(character->yearOfBirth, token);

    token = strsep(&str, ";");
    strcpy(character->eyeColour, token);

    token = strsep(&str, ";");
    strcpy(character->gender, token);

    token = strsep(&str, "\n");
    strcpy(character->hairColour, token);

    // Supondo que se a espécie for "human", então é um bruxo
    character->wizard = (strcmp(character->species, "human") == 0);
}

// Função para imprimir os dados de um personagem
void printCharacter(Character* character) {
    printf("[%s ## %s ## %s ## %s ## %s ## %s ## %s ## %d ## %d ## %s ## %d ## %s ## %s ## %s ## %s ## %s ## %d]\n",
           character->id, character->name, character->alternateNames, character->house, character->ancestry, 
           character->species, character->patronus, character->hogwartsStaff, character->hogwartsStudent,
           character->actor, character->alive, character->dateOfBirth, character->yearOfBirth, character->eyeColour,
           character->gender, character->hairColour, character->wizard);
}

// Adaptação da função readPlayersFromFile para ler os personagens
void readCharactersFromFile(Character* characters[], const char* filename) {
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        perror("Error opening file");
        return;
    }

    char line[1024];
    int i = 0;
    while (fgets(line, 1024, file) && i < 3922) { // 3922 é apenas um exemplo, ajuste conforme necessário
        characters[i] = (Character*) malloc(sizeof(Character));
        if (characters[i] == NULL) {
            perror("Error allocating memory");
            return;
        }
        readCharacterData(line, characters[i]);
        i++;
    }
    fclose(file);
}

// Função para verificar se a entrada é "FIM" para encerrar o programa
bool isEnd(const char *str) {
    return (strcmp(str, "FIM") == 0);
}

int main() {
    const int num_characters = 3922; // ajuste conforme o número real de personagens
    Character* characters[num_characters];

    readCharactersFromFile(characters, "/tmp/characters.csv");

    char input[100];
    scanf("%99s", input);
    while (!isEnd(input)) {
        int position = atoi(input);
        if (position >= 0 && position < num_characters) {
            printCharacter(characters[position]);
        }
        scanf("%99s", input);
    }

    // Libera a memória alocada
    for (int i = 0; i < num_characters; i++) {
        free(characters[i]);
    }

    return 0;
}
